core/src/main/java/jenkins/security/ClassFilterImpl.java

```diff
@@ -0,0 +1,257 @@
+ * yet is not mentioned in {@code whitelisted-classes.txt}.
+ */
+@Restricted(NoExternalUse.class)
+public class ClassFilterImpl extends ClassFilter {
+
+    private static final Logger LOGGER = Logger.getLogger(ClassFilterImpl.class.getName());
+
+    /**
+     * Register this implementation as the default in the system.
+     */
+    public static void register() {
+        if (Main.isUnitTest && Jenkins.class.getProtectionDomain().getCodeSource().getLocation() == null) {
+            mockOff();
+            return;
+        }
+        ClassFilter.setDefault(new ClassFilterImpl());
+    }
+
+    /**
+     * Undo {@link #register}.
+     */
+    public static void unregister() {
+        ClassFilter.setDefault(ClassFilter.STANDARD);
+    }
+
+    private static void mockOff() {
+        LOGGER.warning("Disabling class filtering since we appear to be in a special test environment, perhaps Mockito/PowerMock");
+        ClassFilter.setDefault(ClassFilter.NONE); // even Method on the standard blacklist is going to explode
+    }
+
+    private ClassFilterImpl() {}
+
+    /** Whether a given class is blacklisted. */
+    private final Map<Class<?>, Boolean> cache = Collections.synchronizedMap(new WeakHashMap<>());
+    /** Whether a given code source location is whitelisted. */
+    private final Map<String, Boolean> codeSourceCache = Collections.synchronizedMap(new HashMap<>());
+    /** Names of classes outside Jenkins core or plugins which have a special serial form but are considered safe. */
+    static final Set<String> WHITELISTED_CLASSES;
+    static {
+        try (InputStream is = ClassFilterImpl.class.getResourceAsStream("whitelisted-classes.txt")) {
+            WHITELISTED_CLASSES = ImmutableSet.copyOf(IOUtils.readLines(is, StandardCharsets.UTF_8));
+        } catch (IOException x) {
+            throw new ExceptionInInitializerError(x);
+        }
+    }
+
+    @SuppressWarnings("rawtypes")
+    @Override
+    public boolean isBlacklisted(Class _c) {
+        for (CustomClassFilter f : ExtensionList.lookup(CustomClassFilter.class)) {
+            Boolean r = f.permits(_c);
+            if (r != null) {
+                LOGGER.log(Level.FINER, "{0} specifies a policy for {1}: {2}", new Object[] {f, _c.getName(), r});
+                return !r;
+            }
+        }
+        return cache.computeIfAbsent(_c, c -> {
+            if (ClassFilter.STANDARD.isBlacklisted(c)) { // currently never true: only the name overload is overridden
+                return true;
+            }
+            String name = c.getName();
+            if (Main.isUnitTest && name.contains("$$EnhancerByMockitoWithCGLIB$$")) {
+                mockOff();
+                return false;
+            }
+            if (c.isArray()) {
+                LOGGER.log(Level.FINE, "permitting {0} since it is an array", name);
+                return false;
+            }
+            if (Throwable.class.isAssignableFrom(c)) {
```

Comment:
I am sure `throwables` used to be blocked?  was there not a good reason to block them - has this changed?

Reply:
I do not recall any case of `Throwable`s being (deliberately) blocked. In fact I recall some suggestion to allow `org.codehaus.groovy.runtime.powerassert.PowerAssertionError` and the like through despite their whole package subtree being in the standard blacklist.

I tried for a while to whitelist individual `Throwable`s but found it completely impractical: every miscellaneous kind of exception that might get thrown in the course of a typical Remoting call wound up having to be added. I think this was in `UserResponse`, despite the existence of `ProxyException`, though it has been a while.

Also `ErrorAction` tries to directly record any `Throwable` terminating a Pipeline step, only using `ProxyException` for problematic classes.

In practice `Throwable`s tend to just have a message and stack trace and be unlikely to be a vector for exploits. We can always continue to blacklist any individual class we suspect of being problematic.